src/qibo/abstractions/circuit.py

```diff
@@ -912,3 +912,97 @@ def read_args(args):
                 gate_list[i] = ("M", qubit_list, register)
 
         return len(qubits), gate_list
+
+    def draw(self, line_wrap=None) -> str: # pragma: no cover
+        """Draw text circuit using unicode symbols.
+
+        Args:
+            line_wrap (int): maximum number of characters per line. This option
+                split the circuit text diagram in chunks of line_wrap characters.
+
+        Return:
+            String containing text circuit diagram.
+        """
+        labels = {"h": "H", "x": "X", "y": "Y", "z": "Z",
+                  "rx": "RX", "ry": "RY", "rz": "RZ",
+                  "u1": "U1", "u2": "U2", "u3": "U3",
+                  "cx": "NOT", "swap": "x", "cz": "Z",
+                  "crx": "RX", "cry": "RY", "crz": "RZ",
+                  "cu1": "U1", "cu3": "U3", "ccx": "X",
+                  "id": "I", "collapse": "M", "fsim": "f",
+                  "generalizedfsim": "gf"}
+
+        # build string representation of gates
+        matrix = [[] for _ in range(self.nqubits)]
+        idx = [0] * self.nqubits
+
+        for gate in self.queue:
+            gate_name = labels.get(gate.name)
+            targets = list(gate.target_qubits)
+            controls = list(gate.control_qubits)
+
+            # identify boundaries
+            qubits = targets + controls
+            qubits.sort()
+            min_qubits_id = qubits[0]
+            max_qubits_id = qubits[-1]
+
+            # identify column
+            col = idx[targets[0]] if not controls and len(targets) == 1 else max(idx)
+
+            # extend matrix
+            for iq in range(self.nqubits):
+                matrix[iq] += [ '' for _ in range(1 + col - len(matrix[iq]))]
```

Comment:
I believe the following is slightly more efficient because it doesn't recreate the list:
```suggestion
                matrix[iq].extend((1 + col - len(matrix[iq]))* [''])
```